File: frontend/src/lib/messaging.ts
/**
 * Messaging API types and helpers.
 *
 * Mirrors the backend Pydantic schemas so frontend and backend
 * stay in sync.
 */

import { api } from "@lib/api";

// ---------------------------------------------------------------------------
// Response types (from backend schemas)
// ---------------------------------------------------------------------------

export interface ParticipantResponse {
  user_id: number;
  username: string;
  display_name: string;
  role: string;
  joined_at: string;
}

export interface MessageResponse {
  id: number;
  fhir_communication_id: string;
  sender_id: number;
  sender_username: string;
  sender_display_name: string;
  body: string;
  amends_id: number | null;
  is_amendment: boolean;
  created_at: string;
}

export interface ConversationResponse {
  id: number;
  fhir_conversation_id: string;
  patient_id: string;
  subject: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  participants: ParticipantResponse[];
  last_message_preview: string | null;
  last_message_time: string | null;
  unread_count: number;
  is_participant: boolean;
  can_write: boolean;
  include_patient_as_participant: boolean;
}

export interface ConversationListResponse {
  conversations: ConversationResponse[];
}

export interface ConversationDetailResponse {
  id: number;
  fhir_conversation_id: string;
  patient_id: string;
  subject: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  participants: ParticipantResponse[];
  messages: MessageResponse[];
  is_participant: boolean;
  can_write: boolean;
  include_patient_as_participant: boolean;
}

// ---------------------------------------------------------------------------
// Request types
// ---------------------------------------------------------------------------

export interface CreateConversationRequest {
  patient_id: string;
  subject?: string;
  participant_ids?: number[];
  initial_message: string;
  include_patient_as_participant?: boolean;
}

export interface SendMessageRequest {
  body: string;
  amends_id?: number;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export function fetchConversations(params?: {
  status?: string;
  patient_id?: string;
}): Promise<ConversationListResponse> {
  const qs = new URLSearchParams();
  if (params?.status) qs.set("status", params.status);
  if (params?.patient_id) qs.set("patient_id", params.patient_id);
  const query = qs.toString();
  return api.get<ConversationListResponse>(
    `/conversations${query ? `?${query}` : ""}`,
  );
}

export function fetchConversation(
  conversationId: number,
): Promise<ConversationDetailResponse> {
  return api.get<ConversationDetailResponse>(
    `/conversations/${conversationId}`,
  );
}

export function createConversation(
  data: CreateConversationRequest,
): Promise<ConversationDetailResponse> {
  return api.post<ConversationDetailResponse>("/conversations", data);
}

export function sendMessage(
  conversationId: number,
  data: SendMessageRequest,
): Promise<MessageResponse> {
  return api.post<MessageResponse>(
    `/conversations/${conversationId}/messages`,
    data,
  );
}

export function markConversationRead(
  conversationId: number,
): Promise<{ ok: boolean }> {
  return api.post<{ ok: boolean }>(`/conversations/${conversationId}/read`);
}

export function fetchPatientConversations(
  patientId: string,
  params?: { status?: string },
): Promise<ConversationListResponse> {
  const qs = new URLSearchParams();
  if (params?.status) qs.set("status", params.status);
  const query = qs.toString();
  return api.get<ConversationListResponse>(
    `/patients/${patientId}/conversations${query ? `?${query}` : ""}`,
  );
}

export function joinConversation(
  conversationId: number,
): Promise<ParticipantResponse> {
  return api.post<ParticipantResponse>(`/conversations/${conversationId}/join`);
}

File: backend/app/schemas/messaging.py
"""Pydantic schemas for messaging API endpoints.

Request and response models for conversations, messages, and participants.
"""

from datetime import datetime

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Requests
# ---------------------------------------------------------------------------


class ConversationCreateIn(BaseModel):
    """Create a new conversation about a patient.

    Attributes:
        patient_id: FHIR Patient UUID the conversation is about.
        subject: Optional thread topic / subject line.
        participant_ids: Other user IDs to add as participants.
        initial_message: Body text of the first message.
        include_patient_as_participant: Whether the patient should be
            able to reply to this conversation.
    """

    patient_id: str
    subject: str | None = None
    participant_ids: list[int] = Field(default_factory=list)
    initial_message: str
    include_patient_as_participant: bool = False

    model_config = {"extra": "forbid"}


class MessageCreateIn(BaseModel):
    """Send a message in an existing conversation.

    Attributes:
        body: Message text (supports markdown).
        amends_id: Optional ID of a message this one amends.
    """

    body: str
    amends_id: int | None = None

    model_config = {"extra": "forbid"}


class ConversationStatusUpdateIn(BaseModel):
    """Update a conversation's status.

    Attributes:
        status: New status value.
    """

    status: str = Field(pattern=r"^(new|active|resolved|closed)$")

    model_config = {"extra": "forbid"}


class AddParticipantIn(BaseModel):
    """Tag / add a user to a conversation.

    Attributes:
        user_id: ID of the user to add.
        role: How they are being added.
    """

    user_id: int
    role: str = Field(default="participant", pattern=r"^(participant|tagged)$")

    model_config = {"extra": "forbid"}


class InviteExternalIn(BaseModel):
    """Invite an external user to access a patient's records.

    Attributes:
        email: Recipient email address.
        user_type: ``external_hcp`` or ``patient_advocate``.
    """

    email: str = Field(max_length=255)
    user_type: str = Field(pattern=r"^(external_hcp|patient_advocate)$")

    model_config = {"extra": "forbid"}


class AcceptInviteIn(BaseModel):
    """Accept an invite token (new user registration or existing user grant).

    Attributes:
        token: JWT invite token from the URL.
        username: Username for new user registration (optional if existing).
        password: Password for new user registration (optional if existing).
    """

    token: str
    username: str | None = None
    password: str | None = None

    model_config = {"extra": "forbid"}


# ---------------------------------------------------------------------------
# Responses
# ---------------------------------------------------------------------------


class ParticipantOut(BaseModel):
    """A conversation participant."""

    user_id: int
    username: str
    display_name: str
    role: str
    joined_at: datetime


class MessageOut(BaseModel):
    """A single message in a conversation."""

    id: int
    fhir_communication_id: str
    sender_id: int
    sender_username: str
    sender_display_name: str
    body: str
    amends_id: int | None = None
    is_amendment: bool = False
    created_at: datetime


class ConversationOut(BaseModel):
    """Summary of a conversation (for list views)."""

    id: int
    fhir_conversation_id: str
    patient_id: str
    subject: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime
    participants: list[ParticipantOut]
    last_message_preview: str | None = None
    last_message_time: datetime | None = None
    unread_count: int = 0
    is_participant: bool = False
    can_write: bool = False
    include_patient_as_participant: bool = False


class ConversationListOut(BaseModel):
    """Paginated list of conversations."""

    conversations: list[ConversationOut]


class ConversationDetailOut(BaseModel):
    """Full conversation with all messages."""

    id: int
    fhir_conversation_id: str
    patient_id: str
    subject: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime
    participants: list[ParticipantOut]
    messages: list[MessageOut]
    is_participant: bool = False
    can_write: bool = False
    include_patient_as_participant: bool = False
